File: demo/KSYLiveDemo/KSYPlayerDemo/URLTableViewController.m
//
//  URLTableViewController.m
//  KSYPlayerDemo
//
//  Created by isExist on 16/8/22.
//  Copyright © 2016年 kingsoft. All rights reserved.
//

#import "URLTableViewController.h"
#import "QRViewController.h"

static NSString *kCellWithIdentifier = @"reuseIdentifier";

@interface URLTableViewController ()

@property (nonatomic, strong) NSMutableArray<NSString *> *stringURLs;

@end

@implementation URLTableViewController

- (instancetype)initWithURLs:(NSArray<NSURL *> *)urls {
    if (self = [super init]) {
        _stringURLs = [NSMutableArray array];
        for (NSURL * url in urls) {
            [_stringURLs addObject:[url absoluteString]];
        }
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    self.navigationItem.rightBarButtonItem = self.editButtonItem;
    
    UIBarButtonItem *scanButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(scanQR)];
    self.navigationItem.rightBarButtonItems = @[self.editButtonItem, scanButtonItem];
    
    UIBarButtonItem *cancelButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(cancel)];
    UIBarButtonItem *confirmButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Confirm" style:UIBarButtonItemStyleDone target:self action:@selector(confirm)];
    self.navigationItem.leftBarButtonItems = @[cancelButtonItem, confirmButtonItem];
    
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:kCellWithIdentifier];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)cancel {
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
}

- (void)confirm {
    NSMutableArray<NSURL *> *URLs = [NSMutableArray array];
    for (NSString *stringURL in _stringURLs) {
        [URLs addObject:[NSURL URLWithString:stringURL]];
    }
    [self.navigationController dismissViewControllerAnimated:YES completion:^{
        _getURLs(URLs);
    }];
}

- (void)scanQR {
    QRViewController *qrVC = [[QRViewController alloc] init];
    __weak typeof(qrVC) weakQRVC = qrVC;
    qrVC.getQrCode = ^(NSString *stringQR) {
        [_stringURLs addObject:stringQR];
        typeof(weakQRVC) strongQRVC = weakQRVC;
        [strongQRVC dismissViewControllerAnimated:YES completion:nil];
        [self.tableView reloadData];
    };
    [self presentViewController:qrVC animated:YES completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _stringURLs.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"reuseIdentifier" forIndexPath:indexPath];
    
    // Configure the cell...
    cell.textLabel.text = _stringURLs[indexPath.row];
    
    return cell;
}

// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [_stringURLs removeObjectAtIndex:indexPath.row];
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: demo/KSYLiveDemo_Swift/KSYPlayerDemo_Swift/URLTableViewController.swift
//
//  URLTableViewController.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/13.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class URLTableViewController: UITableViewController {

    let kCellWithIdentifier = "reuseIdentifier"
    
    
    var getURLs: ((_ scannedURLs: [URL]) -> Void)?
    var stringURLs: [String]?
    
    
    init() {
        super.init(style: .plain)
    }
    
    convenience init(urls: [URL]?) {
        self.init()
        
        if let urlArray = urls {
            for url in urlArray {
                stringURLs?.append(url.absoluteString)
            }
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Uncomment the following line to preserve selection between presentations.
        // self.clearsSelectionOnViewWillAppear = NO;
        
        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        navigationItem.rightBarButtonItem = editButtonItem
        
        let scanButtonItem = UIBarButtonItem.init(barButtonSystemItem: .add, target: self, action: #selector(scanQR))
 
        navigationItem.rightBarButtonItems = [editButtonItem, scanButtonItem]
        
        let cancelButtonItem = UIBarButtonItem.init(title: "Cancel", style: .done, target: self, action: #selector(cancel))
        let confirmButtonItem = UIBarButtonItem.init(title: "Confirm", style: .done, target: self, action: #selector(confirm))
        
        navigationItem.leftBarButtonItems = [cancelButtonItem, confirmButtonItem]
        
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: kCellWithIdentifier)
    }

    func cancel() {
        navigationController?.dismiss(animated: true, completion: nil)
    }
    
    func confirm() {
        var URLs: [URL] = Array()
        guard let _ = stringURLs else{
            return
        }
        for urlStr: String in stringURLs! {
            URLs.append(URL.init(string: urlStr)!)
        }
        navigationController?.dismiss(animated: true, completion: { [weak self] in
            if let _ = self?.getURLs {
                self?.getURLs!(URLs)
            }
        })
    }
    
    func scanQR() {
        let qrVC = QRViewController()
        qrVC.getQrCode = { [weak self] (stringQR) -> Void in
            self?.stringURLs?.append(stringQR)
            // TODO: strong self
            self?.dismiss(animated: true, completion: nil)
            self?.tableView.reloadData()
        }
    }
    
    
    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return stringURLs?.count ?? 0
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: kCellWithIdentifier, for: indexPath)

        cell.textLabel?.text = (stringURLs! as NSArray).object(at: indexPath.row) as? String

        return cell
    }

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            // Delete the row from the data source
            tableView.deleteRows(at: [indexPath], with: .fade)
        } else if editingStyle == .insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    
}
